automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/metadata/DataSpec.scala

```diff
@@ -332,4 +338,129 @@ class DataSpec extends FreeSpec with WebBrowserSpec
       }
     }
   }
+
+  "Download should reflect visible columns" - {
+    "no workspace defaults or user preferences" in withWebDriver(downloadPath) { implicit driver =>
+      testFileDownload(
+        initialColumns = List("participant_id", "foo"),
+        expectedColumns = List("participant_id", "foo"))
+    }
+
+    "no workspace defaults, with user preferences" in withWebDriver(downloadPath) { implicit driver =>
+      testFileDownload(
+        initialColumns = List("participant_id", "foo"),
+        userHidden = Some("foo"),
+        expectedColumns = List("participant_id"))
+    }
+
+    "no workspace defaults, with user preferences, new columns" in withWebDriver(downloadPath) { implicit driver =>
+      testFileDownload(
+        initialColumns = List("participant_id", "foo"),
+        userHidden = Some("foo"),
+        importColumns = Some(List("participant_id", "foo", "bar")),
+        expectedColumns = List("participant_id", "bar"))
+    }
+
+    "with workspace defaults, no user preferences" in withWebDriver(downloadPath) { implicit driver =>
+      testFileDownload(
+        initialColumns = List("participant_id", "foo", "bar"),
+        defaultShown = Some(List("participant_id", "foo")),
+        defaultHidden = Some(List("bar")),
+        expectedColumns = List("participant_id", "foo"))
+    }
+
+    "with workspace defaults, no user preferences, new columns" in withWebDriver(downloadPath) { implicit driver =>
+      testFileDownload(
+        initialColumns = List("participant_id", "foo", "bar"),
+        defaultShown = Some(List("participant_id", "foo")),
+        defaultHidden = Some(List("bar")),
+        importColumns = Some(List("participant_id", "foo", "bar", "baz")),
+        expectedColumns = List("participant_id", "foo", "baz"))
+    }
+
+    "with workspace defaults, with user preferences" in withWebDriver(downloadPath) { implicit driver =>
+      testFileDownload(
+        initialColumns = List("participant_id", "foo", "bar"),
+        defaultShown = Some(List("participant_id", "foo")),
+        defaultHidden = Some(List("bar")),
+        userHidden = Some("foo"),
+        expectedColumns = List("participant_id"))
+    }
+
+    "with workspace defaults, with user preferences, new columns" in withWebDriver(downloadPath) { implicit driver =>
+      testFileDownload(
+        initialColumns = List("participant_id", "foo", "bar"),
+        defaultShown = Some(List("participant_id", "foo")),
+        defaultHidden = Some(List("bar")),
+        userHidden = Some("foo"),
+        importColumns = Some(List("participant_id", "foo", "bar", "baz")),
+        expectedColumns = List("participant_id", "baz"))
+    }
+  }
+
+  private def testFileDownload(initialColumns: List[String],
```

Comment:
This is about as complex as I'd like to get with a shared test script. Once it starts doing things conditionally depending on function arguments, I tend to want to back off. However, there was enough still shared about the steps and verifications that, to me, the extra logic pays off in readability of the test cases.